crates/core/src/hash.rs

```diff
@@ -407,3 +426,78 @@ mod keccak {
 }
 
 pub use keccak::Keccak256;
+
+#[cfg(feature = "use_poseidon_halo2")]
+mod poseidon_halo2 {
+    use super::Blinded;
+    use poseidon_halo2::{hash as poseidon_hash, F};
+
+    /// Poseidon hash algorithm with preimage padding.
+    #[derive(Default, Clone)]
+    pub struct PoseidonHalo2 {}
+
+    /// How many least significant bytes of a field element are used to pack the preimage into.
+    const BYTES_PER_FIELD_ELEMENT: usize = 31;
+
+    /// Maximum allowed bytesize of the hash preimage.
+    ///
+    /// If more data than the allowed maximum needs to be hashed, then the data should be split up
+    /// into chunks and each chunk should be hashed separately.
+    pub const POSEIDON_MAX_INPUT_SIZE: usize = 434;
+
+    impl super::HashAlgorithm for PoseidonHalo2 {
+        fn id(&self) -> super::HashAlgId {
+            super::HashAlgId::POSEIDON_HALO2
+        }
+
+        fn hash(&self, _data: &[u8]) -> super::Hash {
+            unimplemented!()
+        }
+
+        fn hash_prefixed(&self, _prefix: &[u8], _data: &[u8]) -> super::Hash {
+            unimplemented!()
+        }
+
+        fn hash_blinded(&self, data: &Blinded<Vec<u8>>) -> super::Hash {
+            let (data, blinder) = data.clone().into_parts();
+
+            let mut field_elements = to_field_elements(&data);
+
+            // Zero-pad to a total of 14 field elements.
+            while field_elements.len() < 14 {
+                field_elements.push(F::zero());
+            }
+            // The last field element is the salt.
+            field_elements.push(to_field_elements(blinder.as_inner())[0]);
+
+            poseidon_hash(&field_elements).into()
+        }
+    }
+
+    fn to_field_elements(data: &[u8]) -> Vec<F> {
+        data.chunks(BYTES_PER_FIELD_ELEMENT)
+            .map(|bytes| {
+                let mut bytes = bytes.to_vec();
+                // Reverse to little-endian.
+                bytes.reverse();
```

Comment:
This is necessary for `F::from_bytes` below which expects LE.

My thinking was that for the devs who will use the hash in their circom circuits, it will be conceptually simpler to think that the 31-byte chunk of plaintext must be viewed as a BE/MSB0 bitstring which represent the field element.
That's why I purposely stick to BE.

Reply:
done